irohad/multi_sig_transactions/state/impl/mst_state.cpp

```diff
@@ -5,15 +5,34 @@
 
 #include "multi_sig_transactions/state/mst_state.hpp"
 
+#include <algorithm>
 #include <utility>
+#include <vector>
 
+#include <boost/algorithm/cxx11/all_of.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/adaptor/transformed.hpp>
 #include <boost/range/algorithm/find.hpp>
 #include <boost/range/combine.hpp>
 #include "common/set.hpp"
 #include "interfaces/iroha_internal/transaction_batch.hpp"
 #include "interfaces/transaction.hpp"
 #include "logger/logger.hpp"
 
+namespace {
+  shared_model::interface::types::TimestampType oldestTimestamp(
+      const iroha::BatchPtr &batch) {
+    auto timestamps =
+        batch->transactions()
+        | boost::adaptors::transformed(
+              +[](const std::shared_ptr<shared_model::interface::Transaction>
+                      &tx) { return tx->createdTime(); });
+    const auto min_it = std::min_element(timestamps.begin(), timestamps.end());
+    assert(min_it != timestamps.end());
+    return min_it == timestamps.end() ? 0 : *min_it;
```

Comment:
It looks like assertion will check only case when batch is empty so maybe better check this explicitly?
Also if empty batches are invalid here may be `optional<TimestampType>` will be a better choice? Technically 0 is a valid value (even though pretty unlikely).

Reply:
A batch cannot have empty transactions by external contract. I added a more explicit assertion to ensure that at the beginning of the function, as you suggested. But I want to keep this one too because I am dereferencing an iterator, which might cause a crash.